substrate/frame/contracts/src/exec.rs

```diff
@@ -1441,11 +1462,53 @@ where
 	}
 
 	fn call_runtime(&self, call: <Self::T as Config>::RuntimeCall) -> DispatchResultWithPostInfo {
-		let mut origin: T::RuntimeOrigin = RawOrigin::Signed(self.address().clone()).into();
+		let mut origin: <T as frame_system::Config>::RuntimeOrigin =
+			RawOrigin::Signed(self.address().clone()).into();
 		origin.add_filter(T::CallFilter::contains);
 		call.dispatch(origin)
 	}
 
+	fn xcm_execute(
+		&self,
+		message: VersionedXcm<CallOf<T>>,
+		max_weight: Weight,
+	) -> DispatchResultWithPostInfo {
+		let origin = RawOrigin::Signed(self.address().clone()).into();
```

Comment:
Is there a custom origin for contracts? As it stands now, a contract cannot be distinguished from an ordinary account, and I feel like this distinction is important in some cases where the recipient really wants to know whether a particular XCM comes from a SC.

Reply:
Yes correct they are generated from the salt and code hash from the contract, and are just regular accounts

https://github.com/paritytech/polkadot-sdk//blob/b2a9f4e647ea5717465ebe67b49d45a0a3c5227a/substrate/frame/contracts/src/lib.rs#L1623-L1630

Reply:
> and I feel like this distinction is important in some cases where the recipient really wants to know whether a particular XCM comes from a SC

You could also make the inverse argument. It might be bad if people could make this distinction as for all intents and purposes a contract should be able act as if it was a plain account. Just that it is governed by code. By allowing actors to discriminate we might hinder innovation.

Reply:
I don't get what you're saying. How is giving the recipient more information about the origin of the XCM a bad thing? Why isn't the recipient able to configure themselves for only receiving XCMs from user accounts and not SCs?

To be clear, I'm really only interested in the XCM origin of the message, i.e. this is currently returning an origin of `Signed(contract_address)`, but I am saying that we should replace it with something like `PalletInstance(contracts_pallet_index)/GeneralKey(contract_address)`.

Reply:
so you are suggesting that we don't use `execute` directly here but instead call `T::XcmExecutor::execute_xcm_in_credit` directly and set a custom origin.
I kind of agree with @athei here though, contracts should be seen as plain accounts even on the chain where the xcm is being executed, The contract has already the ability of dispatching runtime call, and the origin is set likewise, so I don't think we should do thing any differently here

Reply:
> We already have cases where we specifically trust a user to manage its own funds on a destination chain, but not trust the origin chain itself from managing funds on behalf of the user on the destination chain.

Can't the origin chain itself always DescendOrigin to the user's location and manage its funds that way?

Reply:
That was the argument I made before, but it doesn't hold up to scrutiny when you think about the logistics -- how does a chain exactly add in a DescendOrigin to an XCM out of the blue? It must have been using the `send` extrinsic, and if it is, then it also must have been done via Root, which is usually done via referendum. Presumably, no sane governing body would vote to DescendOrigin to a specific account to manage their funds without a really good reason.

Reply:
> Hence, being able to differentiate between user accounts and SCs is important -- it is an issue of trust. I still struggle to see what adverse circumstances that justifies the non-differentiation between user accounts and SCs.

Why would you assign trust based on whether it is a smart contract or not?

> We already have cases where we specifically trust a user to manage its own funds on a destination chain, but not trust the origin chain itself from managing funds on behalf of the user on the destination chain.

A contract can only act on behalf on its own account.

Reply:
Using a smart contract where an account is expected is the main utility of a contract. Let's not remove that.

Reply:
> A contract can only act on behalf on its own account.

This is clearly inferior -- I as a user trust only myself to manage my own account. When I have no given any consent, no other automation can use my funds, such as a chain or an SC, but when I do give consent, I should also have the ability to whitelist some origin representing an SC to handle my funds. Not being able to distinguish between an account and an SC prevents this.

> Using a smart contract where an account is expected is the main utility of a contract.

How did you come up with this principle? This seems to be copying a design flaw from Ethereum, where accounts are treated as first-class citizens. We've explicitly moved away from that model in Substrate and replaced it with origins, so I really don't get how treating SCs as account addresses are the "main utility of a contract".

Reply:
> > A contract can only act on behalf on its own account.
> 
> This is clearly inferior -- I as a user trust only myself to manage my own account. When I have no given any consent, no other automation can use my funds, such as a chain or an SC, but when I do give consent, I should also have the ability to whitelist some origin representing an SC to handle my funds. Not being able to distinguish between an account and an SC prevents this.

I don't really understand. Why should a contract be able to manage **your** funds? Contract addresses and addresses of plain accounts do not overlap.

Reply:
It seems that what we're discussing here is:

- Contracts are primarily used to automate actions on behalf of users. So, the sensible default is to use a contracts account as a regular user account.
- Since Substrate allows us to dispatch calls from a custom origin, we should take advantage of this feature to allow the chain to configure access rules based on the origin when it's appropriate.

I looked into the idea of pallet-contracts dispatching calls with a custom origin. As Keith mentioned,  we could use `pallet::origin` for that.

There's one issue I've encountered with this approach. When using a custom origin, calling `ensure_signed` will fail due to `construct_runtime` generating the `From<RuntimeOrigin> for Result<pallet_contracts::Origin<Runtime, _>>` implementation.

<https://github.com/paritytech/polkadot-sdk/blob/c879d1d5827f1fa4f72621ce185fa08ac3ea8d31/substrate/frame/support/procedural/src/construct_runtime/expand/origin.rs#L287-L296>

Ideally, I would prefer to provide a custom implementation so that our custom origin can be converted into a signed `frame_system::RawOrigin`. Then we could add a custom `type ContractOrigin: EnsureOrigin<Self::RuntimeOrigin>;` to verify whether a dispatched call comes from the Contract origin.

Reply:
So, after discussing with Basti in element, we've found out that there's really no good way of distinguishing automation from real users meaningfully (SC or other kinds of automation can always call OCW and use another origin), hence making a custom origin for SCs doesn't really help.